cloud/cloud_log_controller.cc

```diff
@@ -59,12 +63,50 @@ CloudLogController::~CloudLogController() {
       "[%s] CloudLogController closed.", Name());
 }
 
-std::string CloudLogController::GetCachePath(const Slice& original_pathname) const {
+std::string CloudLogControllerImpl::GetCachePath(const Slice& original_pathname) const {
   const std::string & cache_dir = GetCacheDir();
   return cache_dir + pathsep + basename(original_pathname.ToString());
 }
 
-Status CloudLogController::Apply(const Slice& in) {
+bool CloudLogControllerImpl::ExtractLogRecord(const Slice& input, uint32_t* operation, Slice* filename,
+                                              uint64_t* offset_in_file,
+                                              uint64_t* file_size, Slice* data) {
+  
+  Slice in = input;
+  if (in.size() < 1) {
+    return false;
+  }
+
+  // extract operation
+  if (!GetVarint32(&in, operation)) {
+    return false;
+  }
+  if (*operation == kAppend) {
```

Comment:
if we have the `enum Operation` (check my comment below), we can do `switch` statements here (and make sure compiler will catch it if we add one more op type without handling it